### PR TITLE
[macOS] support secure restorable state by default

### LINKED ISSUE #150062: macOS Sonoma app always logs "WARNING: Secure coding is automatically enabled for restorable state! However, not on all supported macOS versions of this application."
Running app always shows "Secure coding is automatically enabled for restorable state! However, not on all supported macOS versions of this application. Opt-in to secure coding explicitly by implementing NSApplicationDelegate.applicationSupportsSecureRestorableState:."
```
flutter run -d macos
...
Building macOS application...
✓ Built build/macos/Build/Products/Debug/test_create.app
Syncing files to device macOS...                                    39ms
...
The Flutter DevTools debugger and profiler on macOS is available at:
http://127.0.0.1:9101?uri=http://127.0.0.1:55998/3j3JH_UMKmA=/
2024-06-11 12:50:54.702 test_create[25161:2066612] WARNING: Secure coding is automatically enabled for restorable state! However, not on all supported macOS versions of this application. Opt-in to secure coding explicitly by implementing NSApplicationDelegate.applicationSupportsSecureRestorableState:.
```

### PR DESCRIPTION
**Pull request by cbracken**
By default, Flutter apps only do default AppKit app serialisation of Window location etc. and by default, state serialisation in AppKit apps is compatible with `NSSecureCoding`. AppKit apps generated since Xcode 13.2 include this method in the app delegate generated by the default app template.

Background
==========

This method was added to opt into having [de]serialization  require a coder implementing the `NSSecureCoding` protocol. Apple wasn't able to force this across the board, because `NSSecureCoding` limits certain behaviours during deserialisation, which some third-party apps have have previously relied on.

Specific background on the sorts of vulnerabilities that `NSSecureCoding` was designed to prevent are described in the `NSSecureCoding` documentation:
https://developer.apple.com/documentation/foundation/nssecurecoding?language=objc

A demonstration of a root privilege escalation and SIP bypass vulnerability is described in the following blog post: https://sector7.computest.nl/post/2022-08-process-injection-breaking-all-macos-security-layers-with-a-single-vulnerability/

Fixes: https://github.com/flutter/flutter/issues/150062

## Pre-launch Checklist

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [X] I signed the [CLA].
- [X] I listed at least one issue that this PR fixes in the description above.
- [X] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [X] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [X] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

### CODE REVIEW

File: packages/flutter_tools/lib/src/macos/migrations/secure_restorable_state_migration.dart [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+import '../../base/file_system.dart';
+import '../../base/project_migrator.dart';
+import '../../xcode_project.dart';
+
+const String _appDelegateFileBefore = r'''
+class AppDelegate: FlutterAppDelegate {
+  override func applicationShouldTerminateAfterLastWindowClosed(_ sender: NSApplication) -> Bool {
+    return true
+  }
+}''';
```
Comment (loic-sharma) [inline packages/flutter_tools/lib/src/macos/migrations/secure_restorable_state_migration.dart:14]: We won't be able to migrate apps that modified their `AppDelegate`.

Do you think we should log a message recommending the developer overrides  `applicationSupportsSecureRestorableState` if:

1. We find the app delegate file
2. We can't find the string `applicationSupportsSecureRestorableState` in the app delegate file
3. We can't migrate the app delegate file due to modifications

Alternatively, should the tool catch the `Secure coding is automatically enabled for restorable state! However, ...` error and recommend adding `applicationSupportsSecureRestorableState` to the app delegate?

I don't feel strongly about these, feel free to skip.

**Review by loic-sharma — APPROVED** [review]
LGTM!

File: packages/flutter_tools/lib/src/macos/migrations/secure_restorable_state_migration.dart (hunk 1 above)
Comment (cbracken) [reply packages/flutter_tools/lib/src/macos/migrations/secure_restorable_state_migration.dart:14]: I thought about that a bit. My worry is that this log message gets output on every build, so for someone who, for whatever reason really doesn't want to override this, they'll be getting this warning on every build forever.

Perhaps that's no worse than getting the "You didn't override this method" on every single run, forever, from Apple?

Comment (cbracken) [reply packages/flutter_tools/lib/src/macos/migrations/secure_restorable_state_migration.dart:14]: After discussing offline, I'll land this and we can always revert if people complain about it.

**Pull request merged**